pkg/project/project.go

```diff
@@ -28,78 +28,150 @@ const (
 	maxRecursiveIteration = 500
 )
 
-// Detecter is a common interface for project.
-type Detecter interface {
-	Detect() (Result, bool, error)
-	String() string
-}
+// DetectorID represents a detector ID.
+type DetectorID int
 
-// Result contains the result of Detect().
-type Result struct {
-	Project string
-	Branch  string
-	Folder  string
-}
+const (
+	// UnknownDetector is the detector ID used when not detected.
+	UnknownDetector DetectorID = iota
+	// FileDetector is the detector ID for file detector.
+	FileDetector
+	// MapDetector is the detector ID for map detector.
+	MapDetector
+	// GitDetector is the detector ID for git detector.
+	GitDetector
+	// MercurialDetector is the detector ID for mercurial detector.
+	MercurialDetector
+	// SubversionDetector is the detector ID for subversion detector.
+	SubversionDetector
+	// TfvcDetector is the detector ID for tfvc detector.
+	TfvcDetector
+)
 
-// Config contains project detection configurations.
-type Config struct {
-	// Patterns contains the overridden project name per path.
-	MapPatterns []MapPattern
-	// SubmodulePatterns contains the paths to validate for submodules.
-	SubmodulePatterns []regex.Regex
-	// ShouldObfuscateProject determines if the project name should be obfuscated according some rules.
-	ShouldObfuscateProject bool
-}
+const (
+	fileDetectorString       = "project-file-detector"
+	mapDetectorString        = "project-map-detector"
+	gitDetectorString        = "git-detector"
+	mercurialDetectorString  = "mercurial-detector"
+	subversionDetectorString = "svn-detector"
+	tfvcDetectorString       = "tfvc-detector"
+)
 
-// MapPattern contains [projectmap] data.
-type MapPattern struct {
-	// Name is the project name.
-	Name string
-	// Regex is the regular expression for a specific path.
-	Regex regex.Regex
+// String implements fmt.Stringer interface.
+func (d DetectorID) String() string {
+	switch d {
+	case FileDetector:
+		return fileDetectorString
+	case MapDetector:
+		return mapDetectorString
+	case GitDetector:
+		return gitDetectorString
+	case MercurialDetector:
+		return mercurialDetectorString
+	case SubversionDetector:
+		return subversionDetectorString
+	case TfvcDetector:
+		return tfvcDetectorString
+	default:
+		return ""
+	}
 }
 
+type (
+	// Detecter is a common interface for project.
+	Detecter interface {
+		Detect() (Result, bool, error)
+		ID() DetectorID
+	}
+
+	DetecterArg struct {
+		Entity    string
```

Comment:
Let's change this to `Path string` because it's no longer an entity, it can be a project folder or entity path.